graph_djikstra_heap.c

```diff
@@ -0,0 +1,240 @@
+heap keeps array of arrays where each element is int* array[2] = [priority, key]
+*/
+
+typedef struct {
+    int **p_queue;
+    int size;
+} Heap;
+
+Heap* heap_create();
+void heap_free(Heap *);
+
+void heap_push(Heap *, int *);
+int* heap_peek(Heap *);
+int* heap_pop(Heap *);
+void heap_print(Heap *);
+
+static void heap_realloc(Heap *);
+static void heap_heapify(Heap *);
+
+void heap_print(Heap* self) {
+    for(int i = 0; i < self->size; i++) {
+        printf("%d ", self->p_queue[i][0]);
+        printf("%d\n", self->p_queue[i][1]);
+    }
+}
+
+int* heap_peek(Heap *self) {
+    return self->p_queue[0];
+}
+
+int* heap_pop(Heap *self) {
+    if (self->size <=0) {
+        perror("Heap is empty");
+        abort();
+    }
+
+    int* popped_arr = self->p_queue[0];
+    self->size--;
+
+    if (self->size <= 0) {
+        return popped_arr;
+    }
+
+    self->p_queue[0] = self->p_queue[self->size];
+
+    int heap_size = self->size;
+    int** heap_values = self->p_queue;
+
+    int index, left, right, smallest;
+    index = 0;
+
+    //As only one node may be move down, the only cycle is sufficient
+    while(1) {
+        heap_print(self);
+        left = 2 * index + 1;
+        right = 2 * index + 2;
+        smallest = index;
+
+        if (left < heap_size && heap_values[left][0] < heap_values[smallest][0]) {
+            smallest = left;
+        }
+
+        if (right < heap_size && heap_values[right][0] < heap_values[smallest][0]) {
+            smallest = right;
+        }
+
+        if (smallest == index) {
+            break;
+        }
+
+        int *tmp = heap_values[index];
+        heap_values[index] = heap_values[smallest];
+        heap_values[smallest] = tmp;
+
+        index = smallest;
+        }
+
+    return popped_arr;
+}
+
+static void heap_realloc(Heap *self) {
+    int new_heap_size = self->size * 2;
+
+    if(new_heap_size > 100) {
+        new_heap_size = HEAP_MAX_SIZE;
+    }
+
+    int **tmp = realloc(self->p_queue, new_heap_size * sizeof(int *));
+
+    if (tmp == NULL) {
+        perror("Reallocation to increase size failed\n");
+        abort();
+    }
+
+    self->p_queue = tmp;
+
+}
+
+void heap_push(Heap *self, int* priority_key_arr) {
+    if(self->size >= HEAP_MAX_SIZE) {
+        perror("Heap is full\n");
+        abort();
+    }
+
+    if((self->size > 0) && (self->size % HEAP_INIT_SIZE == 0)) {
+        heap_realloc(self);
+    }
+
+    self->p_queue[self->size] = priority_key_arr;
+    self->size++;
+
+    if(self->size > 1) {
+        heap_heapify(self);
+    }
+}
+
+void heap_heapify(Heap *self) {
+    int heap_size = self->size;
+    int** heap_values = self->p_queue;
+
+    int i, index, left, right, smallest;
+
+    // Start from the last non-leaf node to compare it to it's child(ren) and move upwards
+    for (i = (heap_size - 1) / 2; i >= 0; i--) {
+        index = i;
+        // printf("parent index distance val = %d\n", heap_values[index][0]);
+
```

Comment:
bet Lub'uh Gluzznyova's ass, something is wrong here.
look like "Poopa and Loopa case". am i wrong or you messed up some shit here? heapify traversing going top to bottom according my sources..

Reply:
Afaik:
Top to bottom = O(N* log N)
Bottom to top = O(N)